### PR TITLE
Wrong pass  for function constructor (close #571)

### LINKED ISSUE #571: Wrong pass `this` for function constructor
Found in https://testcafe-hhhm.devexpress.com/history?timestamp=1462872434105&siteUrl=http://shareasale.com/

Script for reproduce:
```javascript
    var Ext = {
        toString: function () {
            return 'Ext';
        }
    };
    Ext.data = {
        toString: function () {
            return 'Ext.data';
        }
    };
    Ext.data.Connection = function () {
        console.log(this);
       //Without proxing `this` equal Ext.data.Connection
       //With proxing `this` equal the parent scope - Ext.data
    };
    Ext.data.Connection.toString = function () {
        return 'Ext.data.Connection';
    }

    var obj = new Ext.data.Connection();
```

Possible cause - https://github.com/DevExpress/testcafe-hammerhead/pull/562

### PR DESCRIPTION
**Pull request by churkin**
\cc @miherlosev @VasilyStrelyaev @kirovboris

### CODE REVIEW

**Comment by miherlosev** [comment]
lgtm

File: src/processing/script/transform.js [hunk 1]
```diff
@@ -21,33 +22,60 @@ function getChange (node, parent, key) {
...
-                transform(childNode, node, key, changes, parentChanged);
+                transform(childNode, node, key, changes, state);
         }
     }
 }
 
-export default function transform (node, parent, key, changes, parentChanged) {
-    var nodeChanged = false;
+function isNodeTransformed (node) {
+    return typeof node.originStart !== 'undefined' && typeof node.originEnd !== 'undefined';
```
Comment (inikulin) [inline src/processing/script/transform.js]: Just use `void 0` comparison

File: src/processing/script/transform.js [hunk 2]
```diff
@@ -21,33 +22,60 @@ function getChange (node, parent, key) {
...
+}
+
+function changeFound (state, changes, replacement, parent, key) {
```
Comment (inikulin) [inline src/processing/script/transform.js]: It's not clear what `changeFound` function do. Can you peek better name?

File: src/processing/script/transform.js [hunk 3]
```diff
@@ -21,33 +22,60 @@ function getChange (node, parent, key) {
...
+    if (!state.hasTransformedAncestor) {
```
Comment (inikulin) [inline src/processing/script/transform.js]: Redundant condition, can be combined with nested condition

File: src/processing/script/transform.js [hunk 4]
```diff
@@ -21,33 +22,60 @@ function getChange (node, parent, key) {
     };
 }
 
-function transformChildNodes (node, changes, parentChanged) {
+function createState (currState, node, parent, key, hasTransformedAncestor) {
```
Comment (inikulin) [inline src/processing/script/transform.js]: Whoa, this gets really complicated. Can you elaborate (maybe in comment) how it actually works?

File: src/processing/script/transform.js [hunk 5]
```diff
@@ -21,33 +22,64 @@ function getChange (node, parent, key) {
...
 
-export default function transform (node, parent, key, changes, parentChanged) {
-    var nodeChanged = false;
+function isNodeTransformed (node) {
+    return node.originStart !== void 0 && node.originEnd !== void 0;
+}
 
+function addChangeForTransformedNode (state, changes, replacement, parent, key) {
+    if (!state.hasTransformedAncestor &&
+        (!state.newExpressionAncestor || !isNodeTransformed(state.newExpressionAncestor))) {
```
Comment (inikulin) [inline src/processing/script/transform.js]: Please, avoid such multiline conditions. Keep it single line or even better refactor it to the var with meaningful name

**Comment by churkin** [comment]
FPR

**Comment by inikulin** [comment]
BTW, what is performance impact of this?

**Comment by churkin** [comment]
Slightly reduced for specific cases in which uses complex `new` expressions. For most cases, will not be changed.

**Comment by churkin** [comment]
FPR

**Comment by kirovboris** [comment]
lgtm

**Comment by VasilyStrelyaev** [comment]
lgtm

**Comment by churkin** [comment]
@inikulin ping

**Comment by inikulin** [comment]
lgtm

**Pull request merged**